### ISSUE TITLE
less hangs within infinite loop

### ISSUE THREAD

**Issue by Roslaniec** [gwsw/less#414]
Hi,

Less can enter infinite loop.
Stack trace (version 643):
```
#0  0x0000561c9b4efbb3 in forw_line_seg (curr_pos=-1, skipeol=0, rscroll=0, nochop=1) at input.c:97
#1  0x0000561c9b4fb7ed in get_seg (pos=-1, tpos=189075) at search.c:1165
#2  0x0000561c9b4fbdc9 in search_range (pos=189096, endpos=-1, search_type=16385, matches=0, maxlines=-1, plinepos=0x7ffd62d1e628, pendpos=0x0, plastlinepos=0x7ffd62d1e620) at search.c:1392
#3  0x0000561c9b4fc200 in search (search_type=16385, pattern=0x0, n=1) at search.c:1593
#4  0x0000561c9b4e968a in multi_search (pattern=0x0, n=1, silent=0) at command.c:1105
#5  0x0000561c9b4ea06d in commands () at command.c:1673
#6  0x0000561c9b4e0a23 in main (argc=-1, argv=0x7ffd62d1e838) at main.c:303

```

This bug appeared in **version 581.2**.
**Version 563** is not affected.

Some more explanation about my scenario. I pipe output of my program to less, like:
./myprog | less -Sn
Then I do:
- Search for text, like: **/PATTERN**
- I do _next_ search multiple times, like: **nnnn**
- At some point myprog does not produce any output (finished job, but technically is still running)
- I press Ctrl+C to terminate myprog (it will be stopped indeed)
- But at this moment, less is entering infinite loop as shown above

To my surprise, there are patterns that can be searched for without entering the loop, but some other patterns enter the loop. 
Unfortunately I don't see any obvious rule here. I'm sure it's not pattern as such important here, but rather positions of pattern inside the output.
If I search for a pattern which does not appear at all then everything works fine.
Additionally it seems that if running **less -n**, then problem does not pop up. Looks like flag **-S** is critical here.

Nevertheless, by looking inside the code, it's quite obvious that get_seg() will not quit if pos=-1.

Best Regards,
Karol

**Comment by gwsw** [comment]
When you press ctrl-C, is there a less prompt at the bottom of the screen, or is less paused reading data from the pipe? Does the problem still occur in less-643?

I have not been able to reproduce this. Can you create a test case that reproduces the problem?

search_range is not supposed to allow get_seg to be called with linepos==NULL_POSITION due to the (pos==NULL_POSITION) check in line 1256, but there must be some situation that is not being accounted for.

**Comment by Roslaniec** [comment]
1. after ctrl-c: no visual effects in less
2. less-643: affected
3. I tried to create test case but without success
I will  try to create patch then.

**Comment by Roslaniec** [comment]
The problem is that `forw_line_seg()` can return `NULL_POSISTION`. The trivial commit https://github.com/Roslaniec/less/commit/3b1957a06344bb34ad8ca48de5e809e8d2e03544 will fix the infinite loop issue.

But unfortunately the problem is deeper then this -- or maybe there is simply another issue, not directly related, but just showed up in the same scenario I'm testing here:
- I'm repeating the search with "n" multiple times, until I reach line=1930
- Next time I press "n" _less_ is waiting for filling its buffer (myprog is not producing enough output)
- After pressing Ctrl-C less immediately jumps to line=2077, where it's matching the pattern again.
- There are even two more patterns at lines 2086 and 2093. The output ends at line 2177.

I understand that _less_ was waiting at line 1930 for filling the internal buffer.
But still I see two issues here:
1. After pressing Ctrl-C _less_ should not jump anywhere; it should stay at line 1930; only after pressing "n" should jump to line 2077
2. After pressing Ctrl-C internal state of less is somehow out of sync, because at this moment navigation commands like "UP", "DOWN", "N" are not working! I need to press them a few times until they start working as expected.

The problem I mention here is actually not related to **-S** flag. The same behavior  is visible with _less_ without any flags.
I have quickly checked ver. 563 and it's not perfect as well (works somehow different): after pressing ctrl-c it jumps to other search location, but does not lose internal state sync.
Maybe the new issue could be opened for this problem; or maybe even not, because old _less_ versions were not perfect here as well. 
Anyhow, I don't think it's critical issue.

**Comment by gwsw** [comment]
Thanks, I have merged the PR. I want to do a little more investigation to make sure I fully understand the scenario. I'd also like to investigate the state sync issue you raise, although if it happens without -S it's almost certainly a different issue.

Can you reproduce either or both issues without using your program,  perhaps just by saving the output data and catting it, something like this (to ensure the pipe doesn't get EOF)?
```
( cat output; sleep 10000 ) | less
```

**Comment by Roslaniec** [comment]
OK, with this simple trick I was able to reproduce both problems: infinite loop with old version and sync issue with newest.
Unzip attached file: [test.gz](https://github.com/gwsw/less/files/12389457/test.gz)

`( cat /tmp/test; sleep 10000 ) | less -Sn`
- Search (/) for: `0gQAAC8WAAACBAAALxYAAARtet8`
- Press "n" and keep until you see: Waiting for data...
- Press "n" few more times
- Press: Ctrl-C
- Press: UP arrow (multiple presses required)

PS. I had to randomize original file as it contained sensitive data, therefore if you want to search for other pattern, it must be very short (eg. "kkm" also triggers the bug). In my previous tests not all patterns were triggering the bug.
